### DISCUSSION TITLE
Embedding weights tied to projection out logits

### DISCUSSION THREAD

**Discussion by conceptofmind** [google/flax#2186] [Q&A] [answered]
Hi,

Is there a concise way to implement embedding weights tied to projection out logits?

In PyTorch, we are able to do something like:
```python
def Parallel_Transformer(*, dim, num_tokens, depth, dim_head=64, heads=8, ff_mult=4):

    net = nn.Sequential(
        nn.Embedding(num_tokens, dim),
        *[
            Residual(ParallelTransformerBlock(dim=dim, dim_head=dim_head, heads=heads, ff_mult=ff_mult))
            for _ in range(depth)
        ],
        LayerNorm(dim),
        nn.Linear(dim, num_tokens, bias=False)
    )
    
    net[-1].weight = net[0].weight

    nn.init.normal_(net[0].weight, std=0.02)
    return net
```
I first attempted to do something like this but I was unsure how to use the nn.Embed .attend() method properly. 
```python
class Transformer_model(nn.Module): 
    dim: int 
    num_tokens: int
    depth: int 
    dim_head: int = 64 
    heads: int = 8 
    ff_mult: int = 4
    
    @nn.compact
    def __call__(self, x):
        embed = nn.Embed(self.num_tokens, self.dim, embedding_init = nn.initializers.normal(sttdev=0.02))(x)
        x = ParallelTransformer(dim=self.dim, depth=self.depth, heads=self.heads, dim_head=self.dim_head, ff_mult=self.ff_mult)(embed)
        x = nn.LayerNorm(epsilon = 1e-5, use_bias = False)(x)
        x = nn.Dense(self.num_tokens, use_bias = False)(x)
        out = embed.attend(x)
        return out
```
This seems to work? But it is not very clean and concise.
```python
class Transformer_model(nn.Module): 
    dim: int 
    num_tokens: int
    depth: int 
    dim_head: int = 64 
    heads: int = 8 
    ff_mult: int = 4    

    @nn.compact
    def __call__(self, x):
        embedding = (self.param('embedding', nn.initializers.normal(), (self.num_tokens, self.dim)) * 0.02)
        x = embedding[x]
        x = ParallelTransformer(dim=self.dim, depth=self.depth, heads=self.heads, dim_head=self.dim_head, ff_mult=self.ff_mult)(x)
        x = nn.LayerNorm(epsilon = 1e-5, use_bias = False)(x)
        x = jnp.dot(x, embedding.T)
        return x
```

Any advice would be greatly appreciated.

Thank you,

Enrico

**Comment by marcvanzee** [comment, marked as answer]
(*Updated Jun 14:* Fixed typo -- see reply @conceptofmind below)

Using `attend` seems like the right approach, but in your example you are not defining `embed` correctly: it is assigned the output of embedding `x`. Also, I think you should get rid of the last `Dense` layer (you are replacing this with `attend`).

So with those changes the code looks as follows:

```py
class Transformer_model(nn.Module): 
  dim: int 
  num_tokens: int
  depth: int 
  dim_head: int = 64 
  heads: int = 8 
  ff_mult: int = 4
  
  @nn.compact
  def __call__(self, x):
    embed = nn.Embed(self.num_tokens, self.dim, embedding_init = nn.initializers.normal(stddev=0.02))
    x = embed(x)
    x = ParallelTransformer(dim=self.dim, depth=self.depth, heads=self.heads, dim_head=self.dim_head, ff_mult=self.ff_mult)(x)
    x = nn.LayerNorm(epsilon = 1e-5, use_bias = False)(x)
    out = embed.attend(x)
    return out
```

**Reply by conceptofmind** [reply]
Hi @marcvanzee ,

I greatly appreciate your help. 

I made the adjustments that you recommended. I had one additional question.

Should the line below be taking embed as the input? It throws a type error for me.
```python
x = ParallelTransformer(dim=self.dim, depth=self.depth, heads=self.heads, dim_head=self.dim_head, ff_mult=self.ff_mult)(embed)
```
When I use x as the input instead it seems to work appropriately.
```python
x = ParallelTransformer(dim=self.dim, depth=self.depth, heads=self.heads, dim_head=self.dim_head, ff_mult=self.ff_mult)(x)
```
So the full code segment would look like:
```python
class Transformer_model(nn.Module): 
  dim: int 
  num_tokens: int
  depth: int 
  dim_head: int = 64 
  heads: int = 8 
  ff_mult: int = 4
  
  @nn.compact
  def __call__(self, x):
    embed = nn.Embed(self.num_tokens, self.dim, embedding_init = nn.initializers.normal(stddev=0.02))
    x = embed(x)
    x = ParallelTransformer(dim=self.dim, depth=self.depth, heads=self.heads, dim_head=self.dim_head, ff_mult=self.ff_mult)(x)
    x = nn.LayerNorm(epsilon = 1e-5, use_bias = False)(x)
    out = embed.attend(x)
    return out
```

Thank you,

Enrico

**Reply by marcvanzee** [reply]
Ah yes you are right, sorry that was a typo in my code. You indeed want to pass the embedding of the input to `ParallelTransformer` and not the embedding layer itself.

**Reply by marcvanzee** [reply]
I fixed the typo in my original reply so other users won't be confused.

**Reply by conceptofmind** [reply]
Hi @marcvanzee ,

Thank you for the help and response. 

I had wanted to confirm that I was not doing something incorrectly with passing the embedding input. 

Once I have finished the data loading and training script would you be willing to check a Pull Request for the full training example in JAX/Flax?

Thank you,

Enrico

**Reply by marcvanzee** [reply]
We aim to keep the number of examples in our main repo relatively small, because we want to make sure we can maintain them. Therefore, we generally recommend to add a new [Show & Tell Discussion thread](https://github.com/google/flax/discussions/categories/show-and-tell) when you have finished an implementation, and we will make sure we link to it from our [Community Examples](https://github.com/google/flax/tree/main/examples#community-examples) page.

**Reply by conceptofmind** [reply]
Hi @marcvanzee ,

I will add a new Show & Tell Discussion thread once the implementation I completed.

Thank you,

Enrico Shippole
